renderdoc/driver/shaders/dxbc/dxbc_debug.h

```diff
@@ -147,6 +147,33 @@ struct GlobalState
   std::map<SampleEvalCacheKey, ShaderVariable> sampleEvalCache;
 };
 
+class DebugAPIWrapper
+{
+public:
+  virtual void SetCurrentInstruction(uint32_t instruction) = 0;
+  virtual void AddDebugMessage(MessageCategory c, MessageSeverity sv, MessageSource src,
+                               std::string d) = 0;
+
+  virtual bool CalculateMathIntrinsic(DXBC::OpcodeType opcode, const ShaderVariable &input,
+                                      ShaderVariable &output1, ShaderVariable &output2) = 0;
+
+  virtual ShaderVariable GetSampleInfo(DXBC::OperandType type, bool isAbsoluteResource, UINT slot,
+                                       const char *opString) = 0;
+
+  virtual ShaderVariable GetBufferInfo(DXBC::OperandType type, UINT slot, const char *opString) = 0;
+  virtual ShaderVariable GetResourceInfo(DXBC::OperandType type, UINT slot, uint32_t mipLevel,
+                                         int &dim) = 0;
+
+  virtual bool CalculateSampleGather(DXBC::OpcodeType opcode, DXBC::ResourceDimension resourceDim,
+                                     DXBC::ResourceRetType resourceRetType, int resourceSampleCount,
+                                     DXBC::SamplerMode samplerMode, ShaderVariable uv,
+                                     ShaderVariable ddxCalc, ShaderVariable ddyCalc, UINT texSlot,
+                                     UINT samplerSlot, float samplerBias, const int texelOffsets[3],
+                                     int multisampleIndex, float lodOrCompareValue,
+                                     const char *swizzle, const char *gatherChannel,
+                                     const char *opString, ShaderVariable &output) = 0;
```

Comment:
I tend not to be strict about this kind of thing but I think this function signature is a bit too large to have everything loose. It doesn't have to be all converted into a single struct, but at least you could gather texture parameters: `resourceDim`, `resourceRetType`, `resourceSampleCount`, `texSlot`, sampler parameters: `samplerMode`, `samplerSlot` and maybe also gather the sampling/co-ordinate parameters (though after combining the above it might be reasonable).

Also a slight nit-pick, but I'd prefer to pass the swizzle as `int[4]` with 0,1,2,3 values, and gatherChannel as an int with 0,1,2,3, then do the conversion to string on the other side. It's only a couple of lines/switch but I think it's better placed on the D3D side of things. The DXBC side can still convert 0xff for unused swizzles into 0 in the array.

Reply:
Bundled the resource and sampler data into single params for each, and moved the string creation of swizzle/channel to the D3D11 side.